### PR TITLE
Word piece

### PR DESCRIPTION
**Pull request by mattdangerw**
Adds a base class for tokenizers and a layerized WordPiece wrapping tf.text.

Will add tests shortly, opening to have a diff to comment on for API design questions.

### CODE REVIEW

File: keras_nlp/layers/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            inputs: Input tensor, or dict/list/tuple of input tensors.
+            *args: Additional positional arguments.
+            **kwargs: Additional keyword arguments.
+        """
+        raise NotImplementedError(
+            f"{self.__class__.__name__}.detokenize is not implemented."
+            "This class may not support reversing tokenization."
+        )
+
+    def call(self, inputs: tf.Tensor, *args, **kwargs):
```
Comment (mattdangerw) [inline keras_nlp/layers/tokenizer.py:57]: discussion notes: rout everything though __call__ with flags! but as an end user still call tokenize/detokenize

**Comment by fchollet** [comment]
The main design question is probably this:

- Using the "raw" `tokenize`/`detokenize` methods is fine in tf.data, but when building a model with the functional API, you need to be using `__call__()`
- Hence, we can make `__call__()` support all possible modes (with routing to one mode or another via keyword arguments), and have `tokenize`/`detokenize` defer to `__call__()` with the appropriate mode arg
- However, this raises the question of what the layer implementer is supposed to override

Some options:

**Option A**: have new implementer-facing methods `tokenize_impl()` / `detokenize_impl()` or similar which will be called by the base class `call()` method. This mirrors a common pattern where this type of situation is handled by having the implementer override a method that's separate from the interface method, which wraps the overridden method. For instance, when writing layers, you implement `call()` but not `__call__()`.

**Option B**: have instantiation-time wrapping of `tokenize()` / `detokenize()`. It's like option A but where the overridden `tokenize` gets replaced with a wrapped version of itself upon instantiation. It works but is somewhat dirty. We do this on the `Metric` base class with `update_state()`.

**Option C**: have implementers implement `call(self, ..., mode="tokenize")` with the convention that they should do `if mode == "tokenize": ...` / `if mode == "detokenize": ...`

I think the simplest is probably option C. WDYT?

**Comment by mattdangerw** [comment]
Thanks! Agreed on the framing. I am not immediately sure on best solution.

We might have a few more things we need to cover as the base class evolves. Tf text has a `tokenizer_with_offsets` which gives byte offset into the original text (useful for things like named entity recognition I believe). If we start to have a few different flags we rout through call things might look a little more chaotic on the subclass. E.g.

```
class MyTokenizer:
  def call(self, inputs, detokenize=False, add_offsets=False):
    if detokenize:
      return some_thing
    if add_offsets:
      return other_thing
    if detokenize && add_offsets:
      raise "both not supported"
    return main_tokenization
```

So I might have a slight preference for **Option A**. It would reduce the need to replicate branching logic on the subclass, and allow us to default unhanded methods which an exception. That could be useful for a broader ecosystem. We add some new method down the line to the base class, and all existing tokenizers will correctly raise a `NotImplementError` until they update to support.

**Comment by mattdangerw** [comment]
I guess I was missing part of your proposal for **Option C**, we could follow a call block like:

```
  def call(self, inputs, mode="tokenize"):
    if mode == "tokenize":
      ...
    elif mode == "tokenize_with_offsets":
      ...
    elif mode == "detokenize":
      ...
    else:
      raise NotImplementedError
```

Given that, I'm probably not very opinionated between options A and C.

**Comment by mattdangerw** [comment]
Sounds good to me! Thanks for the input here. I'll finish up this PR with option C and we can see how it looks.

**Comment by mattdangerw** [comment]
Another thought, it would be good to be able to document the tokenize, detokenize methods for each class, with more details on how they operate per tokenizer. We did this for `adapt` on preprocessing layers by just redefining the method and chaining to super(), though that's another slightly awkward point for subclassers. Probably no need to worry about this until we start generating documentation, but something to think about.

**Comment by fchollet** [comment]
Perhaps we could simply require folks to add relevant sections in the class docstrings? Layer docs are typically generated purely from class docstrings.

**Comment by mattdangerw** [comment]
> Layer docs are typically generated purely from class docstrings.

Not for the tf docs (e.g. [adapt](https://www.tensorflow.org/api_docs/python/tf/keras/layers/StringLookup#adapt)), which I guess is why we did that workaround. But seems fine to do here.

**Comment by mattdangerw** [comment]
Okay, played around more with this and we now think **Option B** may be the best bet. **Option C** became cumbersome when subclassing a subclass of Tokenizer, when it was not easy to extend the tokenize and detokenize methods without redefining and chaining to the private helpers within call.

**Option B** is probably the nicest for end users. Who can just implement the same method that they call on instantiated objects, and not need to worry about functional models in the mental model at all.

A version of **Option B** base class is now on this PR.

**Comment by mattdangerw** [comment]
I have also tried moving these to a separate `keras_nlp.tokenzers.SomeTokenizer` namespace. Upon thinking about it more, I think there will be an expectation to have tokenizers grouped together from other similar APIs, and overall I think it will bring more clarity.

Since this is a custom base class, and users will not be implementing layers in the normal sense (the expectation is now to not implement call), I think this is a reasonable naming choice.

File: keras_nlp/tokenizers/tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    def tokenize(self, inputs: tf.Tensor, *args, **kwargs) -> tf.Tensor:
+        """Transform input tensors of strings into output tokens.
+
+        Args:
+            inputs: Input tensor, or dict/list/tuple of input tensors.
+            *args: Additional positional arguments.
+            **kwargs: Additional keyword arguments.
+        """
+        raise NotImplementedError(
+            "No implementation of `tokenize` was found for "
```
Comment (fchollet) [inline keras_nlp/tokenizers/tokenizer.py:65]: Nit: refer to methods/functions with a `()` at the end to make it clear they're methods/functions

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    argument, which should be either an integer or string type.
+
+    Args:
+        vocabulary: A list of strings or a string string filename path. If
+            passing a list, each element of the list should be a single word
+            piece token string. If passing a filename, the file should be a
+            plain text file containing a single word piece token per line.
+        sequence_length: If set, the output will be converted to a dense
+            tensor and padded/trimmed so all outputs are of sequence_length.
+        lower_case: If true, the input text will be first lowered before
```
Comment (fchollet) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:100]: Should this be `lowercase`? "lowercase" is more common and is what we use when describing this behavior in `TextVectorization`.

**Review by fchollet — APPROVED** [review]
Thank you!

File: keras_nlp/tokenizers/tokenizer.py (hunk 2 above)
Comment (mattdangerw) [reply keras_nlp/tokenizers/tokenizer.py]: Done

File: keras_nlp/tokenizers/word_piece_tokenizer.py (hunk 3 above)
Comment (mattdangerw) [reply keras_nlp/tokenizers/word_piece_tokenizer.py]: Yeah, I think that makes sense. Most of the BERT stuff out there seems to use `do_lower_case` which I think is picked up from the [original BERT implementation](https://github.com/google-research/bert/blob/master/tokenization.py#L164).

But it seems lowercase is the more correct usage https://english.stackexchange.com/questions/59409/lowercase-lower-case-or-lower-case

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+            `split_pattern` of delimeters that should be kept as independent
+            tokens. By default, all punctuation marks will be kept as tokens.
+        suffix_indicator: The characters prepended to a wordpiece to indicate
+            that it is a suffix to another subword.
+        oov_token: The string value to substitute for an unknown token. It
+            must be included in the vocab.
+
+    References:
+        [Schuster and Nakajima, 2012](https://research.google/pubs/pub37842/)
+        [Song et al., 2020](https://arxiv.org/abs/2012.15524)
```
Comment (fchollet) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:118]: Nit: please use markdown list

Comment (mattdangerw) [reply keras_nlp/tokenizers/word_piece_tokenizer.py:118]: done

**Review by chenmoneygithub — APPROVED** [review]
